File: src/main/java/DecoratorAtaques.java
public abstract class DecoratorAtaques implements Ataque {
    private AtaqueConcret decorado;

    DecoratorAtaques(AtaqueConcret aDecorar) {
        decorado = aDecorar;
    }
    public String getNombre() {
        return decorado.getNombre();
    }
    public int getDanio() {
        return decorado.getDanio();
    }
    public int getRecupera() {
        return decorado.getRecupera();
    }
    public int getVentaja()  {
        return decorado.getVentaja();
    }

    public void realizarAtaque(Chinpokomon atacado, Chinpokomon atacante) {
        if (this.tieneEfecto()){
            atacado.modificarVida( - this.especialEfecto(atacado) );
        }else{
            decorado.realizarAtaque(atacado,atacante);
        }
    }
    public boolean tieneVentaja(Chinpokomon atacado, Chinpokomon atacante) {
        return decorado.tieneVentaja(atacado,atacante);
    }
    public boolean tieneEfecto() { return false;}
    int especialEfecto(Chinpokomon atacado) {
        return -1;
    }
}

File: python/DecoratorAtaques.py
from AtaqueConcrete import Ataque


class DecoratorAtaques (Ataque) :
    def __init__(self, aDecorar):
        self.__decorado = aDecorar
    
    def get_nombre(self):
        return self.__decorado.get_nombre()
    def get_danio(self):
        return self.__decorado.get_danio()
    def get_recupera(self):
        return self.__decorado.get_recupera()
    def get_ventaja(self):
        return self.__decorado.get_ventaja()
    
    def realizar_ataque(self, atacado, atacante):
        if (self.tiene_efecto()):
            atacado.modificar_vida( - self.especial_efecto(atacado) )
        else:
            self.__decorado.realizar_ataque(atacado,atacante)
    
    def tiene_ventaja(self, atacado, atacante):
        return self.__decorado.tiene_ventaja(atacado,atacante)
    
    def tiene_efecto(self):
        return False
    def especial_efecto(self, atacado):
        return -1